src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,163 @@
+            lastNode = firstNode;
+        } else {
+            lastNode.next = new Node<>(lastNode, value, null);
+            lastNode = lastNode.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, firstNode);
+            firstNode.prev = newNode;
+            firstNode = newNode;
+            size++;
+        } else {
+            Node<T> afterNode = getNode(index);
+            Node<T> beforeNode = afterNode.prev;
+            Node<T> newNode = new Node<T>(beforeNode, value, afterNode);
+            afterNode.prev = newNode;
+            if (beforeNode != null) {
+                beforeNode.next = newNode;
+            } else {
+                firstNode = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = getNode(index).item;
+        getNode(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> removed = getNode(index);
+        T value = removed.item;
+        unlink(removed);
+        return value;
+
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = getNode(object);
+        if (node == null) {
+            return false;
+        }
+        unlink(node);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < 0) {
+            throw new IndexOutOfBoundsException("Can't get element by negative index");
+        } else if (index >= size) {
+            throw new IndexOutOfBoundsException("No such element exists");
+        } else if (index < size / 2) {
+            Node<T> sought = firstNode;
+            for (int i = 1; i <= index; i++) {
+                sought = sought.next;
+            }
+            return sought;
+        } else {
+            Node<T> sought = lastNode;
+            for (int i = size - 2; i >= index; i--) {
+                sought = sought.prev;
+            }
+            return sought;
+        }
+    }
+
+    private Node<T> getNode(T value) {
+        Node<T> node = firstNode;
+        for (int i = 0; i < size; i++) {
+            if (Objects.equals(node.item, value)) {
```

Comment:
we shouldn't use class Object.

Reply:
but it's class Objects, not Object

Reply:
Objects* . It's written in Common mistakes )